### PR TITLE
[5.9] Fix SILCombine to preserve a release of a move-only type

### PR DESCRIPTION
**Pull request by atrick**
The release needs to be preserved in case a user-defined deinit is present in the released type. Checking for move-only is slightly conservative.

Fixes rdar://109846094 ([move-only] SILCombine eliminates struct deinitialization)

(cherry picked from commit dc974e8c00d923474905d9c41d6b71bc045751d9)

- Explanation: SILCombine removes releases of aggregates whose members don't contain any references. This deletes struct deinitializers, so they silently never run.

- Scope of Issue: A struct's deinit would silently fail to run when the struct was a member within another type. In the most trivial cases, with no type composition, and earlier MoveOnlyDeinitInsertion pass was hiding the problem.

- Risk: Zero risk to regular copyable code. And non-copyable was already broken.

- PR to main: https://github.com/apple/swift/pull/66161

- Reviewed By: Michael Gottesman

- Resolves: rdar://109846094 ([move-only] SILCombine eliminates struct deinitialization)

### CODE REVIEW

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
@gottesmm please merge after your review

**Review by gottesmm — APPROVED** [review]
Thanks!

**Comment by atrick** [comment]
@tbkka waiting for your approval

**Review by tbkka — APPROVED** [review]

**Pull request merged**
